### PR TITLE
Js2 - What to do with profile page?

### PR DESCRIPTION
**Pull request by Ramsnes**
- What are your suggestions to do with the profile page?
- Sorting posts with first/latests doesn't work. Only A-Z/Z-A
- I was wondering to either try and bring the logged in username and posts upon login.
Or maybe even delete the profile page if I don't manage to create codes for it. I now have the CSS frameworks contents there as default which will be defeated.

### CODE REVIEW

File: index.html [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+          </li>
+          <li class="nav-item">
+            <a href="profile/index.html" class="nav-link px-2">Profile</a>
+          </li>
+        </ul> -->
+        <p class="text-center">© 2023 CodeSharing, Inc</p>
+      </footer>
+      <!-- Footer end -->
+    </main>
+    <script type="module" src="../js/login.js"></script>
```
Comment (cnnrbrn) [inline index.html]: The path here is incorrect, ../ means go up one directory, use a single .

File: js/login.js [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+        body: postBody,
+      },
+      false
+    );
+    const token = userLoginData.accessToken;
+    addToLocalStorage("accessToken", token);
+
+    window.location.href = "/profile/index.html";
+  } catch (error) {
+    console.error("Login failes", error);
```
Comment (cnnrbrn) [inline js/login.js]: You need to display a message to the user in the html here, not just log the error. The user will never check the console for the error message.

Very basic example:

`document.querySelector("#message").innerHTML = '<div class="error">Some error message</div>';`

Obviously you would need an element with id="message" in the html.

File: js/fetcher.js [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+        ...fetchOptions,
+        headers: {
+          ...fetchOptions.headers,
+          Authorization: `Bearer ${accessToken}`,
+        },
+      };
+    }
+
+    const response = await fetch(url, fetchOptions);
+    const data = await response.json();
```
Comment (cnnrbrn) [inline js/fetcher.js:30]: I would check if the request was a success before returning the results/data. You can use the ok property to check, for example:

```js
if(response.ok) {
    return data;
 }
 // response not ok, throw an error, catch it in the try/catch of the calling function
 throw new Error(data.errors[0].message);
```

File: js/fetcher.js [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+    return data;
+  } catch (error) {
+    console.error("Error with API call", error);
```
Comment (cnnrbrn) [inline js/fetcher.js:33]: If you catch the error and just log it, the try/catch where you call the function (like in loginUser function, for example), will never receive the error in the catch block. I would remove the try/catch in this function.

File: js/register.js [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+
+const form = document.querySelector("#registrationForm");
+const name = document.querySelector("#registerUsername");
+const email = document.querySelector("#registerEmail");
+const password = document.querySelector("#registerPassword");
+
+async function registerUser(user) {
+  console.log("Register user:", user); //User = user object
+  // POST request
+  const postBody = JSON.stringify(user);
```
Comment (cnnrbrn) [inline js/register.js:13]: The call to fetcher should be in a try/catch block like in loginUser. Display a message to the user (in html) if the request was a success or a failure.

File: js/posts.js [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+    body: postBody,
+    media: postMedia,
+    created: new Date().toISOString(),
+    updated: new Date().toISOString(),
+    tags: tags.split(","), // Splits the tags strong into an array
+  };
+
+  try {
+    const newPost = await fetcher(
+      "https://api.noroff.dev/api/v1/social/posts",
```
Comment (cnnrbrn) [inline js/posts.js]: The url should be imported from the constants file, not hard-coded here.

File: js/posts.js [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+      },
+      true
+    );
+
+    // Render of new post on the page
+    renderPost(newPost);
+
+    console.log("New post created:", newPost);
+  } catch (error) {
+    console.error("Error creating a new post:", error);
```
Comment (cnnrbrn) [inline js/posts.js:49]: See the comment in login.js. Always display a message to the user in html.

File: js/postDetails.js [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+import { BASE_API_URL } from "./common/constants.js";
+
+document.addEventListener("DOMContentLoaded", () => {
+  // Get post ID from the URL
+  const postId = new URLSearchParams(window.location.search).get("id");
+
+  // Fetch post details
+  const apiUrl = `${BASE_API_URL}/social/posts/${postId}`;
+
+  fetcher(apiUrl, { method: "GET" }, true)
```
Comment (cnnrbrn) [inline js/postDetails.js:12]: It's better to be consistent in the way you make API calls. Here you are using then(), but in other files you are using async/await. Stick to one way, the one that makes most sense to you.

**Comment by cnnrbrn** [comment]
For the profile page, you can get the profile of the logged in user: `/social/profiles/<name>` https://docs.noroff.dev/social-endpoints/profiles#single-entry

When you login, you are currently only storing the access token. You can also store the other user details, or just the name, then get the name from storage on the profile page, add it to the URL like above, and make the API call to get the profile details.

File: feed/index.html [hunk 9]
```diff
@@ -0,0 +1,200 @@
...
+      <!-- Footer end -->
+    </main>
+    <!-- Enables sorting text  -->
+    <script
+      src="https://cdn.jsdelivr.net/npm/bootstrap@5.0.0-beta1/dist/js/bootstrap.bundle.min.js"
+      integrity="sha384-ygbV9kiqUc6oa4msXn9868pTtWMgiQaeYH7/t7LECLbyPA2x65Kgf80OJFdroafW"
+      crossorigin="anonymous"
+    ></script>
+    <script type="module" src="../js/logout.js"></script>
+    <script type="module" src="../js/fetcher.js"></script>
```
Comment (cnnrbrn) [inline feed/index.html]: If you are using modules (imports/exports), which you are, you don't need to link to the files here. For example, you import and use fetcher in posts.js, you don't need to link to fetcher.js like this. The modules handle that.

File: js/posts.js
Comment (cnnrbrn) [inline js/posts.js:1]: There are no posts listed on this page by default, only after searching by tag.

You could fetch and display the logged-in user's posts here using: `/social/profiles/<name>/posts `
https://docs.noroff.dev/social-endpoints/profiles#all-posts-by-profile

See the comment about storage the name when logging in.

File: js/common/utils/sort.js [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+});
+
+function sortPosts(sortBy) {
+  const postsContainer = document.getElementById("postsContainer");
+  const posts = Array.from(postsContainer.children);
+  // case names appended to id's in feed html
+  switch (sortBy) {
+    case "latest":
+      posts.sort((a, b) => {
+        const dateA = new Date(a.getAttribute("data-created"));
```
Comment (cnnrbrn) [inline js/common/utils/sort.js:26]: I can't see the `date-created` attribute in the HTML?

Comment (Ramsnes) [reply js/common/utils/sort.js:26]: > I can't see the `date-created` attribute in the HTML?

No I sadly havent figured out how to do this. I'm struggling to implement this properly as I don't truly understand it.

File: js/posts.js
Comment (Ramsnes) [reply js/posts.js:1]: I'm trying to just fetch and render the 10 latest posts from the server on feed/index.html, but I am also struggling here with that. Every time I try, something else breaks. So I don't know how.

**Pull request closed without merging**